### ISSUE TITLE
(Resolved) 6x Slower than FluxGYM?

### ISSUE THREAD

**Issue by plugcrypt** [kijai/ComfyUI-FluxTrainer#51]
For the same training settings on both trainer:
512x512
Adamw8bit
4 batch size

FluxGym : 1.19s/it
ComfyUI FluxTrainer: 7.31s/it

About 6 times faster on FluxGym, on FluxTrainer my gpu utilisation is going up and down as well with my power draw, the vram usage doen't go above 55% with this training setting, I run this on a 4090

I hope that I'm missing something

Using these arguments : --use-pytorch-cross-attention --fast --highvram --dont-upcast-attention 
Eveything up to date, comfyUI + dependencies

**Comment by kijai** [comment]
Pytorch version? You'll want 2.4.1.

Those are definitely not normal speeds.

**Comment by plugcrypt** [comment]
![image](https://github.com/user-attachments/assets/b8afeb93-96a7-4833-9dd1-b6121275628e)

**Comment by kijai** [comment]
Can you show the settings you are using?

**Comment by plugcrypt** [comment]
Sure, 
![workflow (29)](https://github.com/user-attachments/assets/39154b1b-059c-48f3-9a99-c1865344e27d)

**Comment by kijai** [comment]
Split_mode is a memory saving feature that slows it down a lot as it offloads half of the mode tom RAM and back constantly.

**Comment by plugcrypt** [comment]
I got  3.94s/it after a 10min of training, so it's definitely faster, and my power draw stays around 410w, it dosen't spike as before, nice. I would like to how what settings are set on FluxGym to get around 1.2s/it, i'm still searching how to update it to edit the script before the run, https://x.com/cocktailpeanut/status/1832787205774786710 I didn't find this update button yet

**Comment by plugcrypt** [comment]
This is my iteration rate on FluxGym atm
![image](https://github.com/user-attachments/assets/c3fb6c0c-37e2-44b1-907a-068d9447493d)

**Comment by kijai** [comment]
> This is my iteration rate on FluxGym atm ![image](https://private-user-images.githubusercontent.com/88064700/366225995-c3fb6c0c-37e2-44b1-907a-068d9447493d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMDY5NDAsIm5iZiI6MTcyNjAwNjY0MCwicGF0aCI6Ii84ODA2NDcwMC8zNjYyMjU5OTUtYzNmYjZjMGMtMzdlMi00NGIxLTkwN2EtMDY4ZDk0NDc0OTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEwVDIyMTcyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlNGFkNWQzNDYxOGJhNDFjYjAyZWIyYmMxYjJjYjI5NGM4MTNmODdkMWFlNDM2OTU4MjRiOThiNTQwMzcwYzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OJOXVICifGLkqOJv6Pq_WxbnlhFoEdaidcymsb8waV8)

Are you absolutely certain that's using batch_size 4? Seems too fast for that, what would the speed be with batch_size 1?

**Comment by plugcrypt** [comment]
No it's a mistake I made on my first post, But I previously compaired both with batch size 1 with the same settings and noticed a huge difference 6s/it on comfy, 1.2s/it on flux, but I didn't tried to compair this again with the split mode to false, and yes both are running at the same speed, this was the culprit. Thank you for the help appreciate

![image](https://github.com/user-attachments/assets/758b1ae7-59b0-4382-a4de-007a629a100f)

**Comment by plugcrypt** [comment]
At the end, I thought that I could increase the batch size until I reach the vram limit, I didn't knew that this could have such an impact on the speed of the training. Since I'm moslty working on styles I think I will keep using 3 or 4 batch size, if it's making sense

**Comment by kijai** [comment]
> No it's a mistake I made on my first post, But I previously compaired both with batch size 1 with the same settings and noticed a huge difference 6s/it on comfy, 1.2s/it on flux, but I didn't tried to compair this again with the split mode to false, and yes both are running at the same speed, this was the culprit. Thank you for the help appreciate
> 
> ![image](https://private-user-images.githubusercontent.com/88064700/366238064-758b1ae7-59b0-4382-a4de-007a629a100f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMTAwODUsIm5iZiI6MTcyNjAwOTc4NSwicGF0aCI6Ii84ODA2NDcwMC8zNjYyMzgwNjQtNzU4YjFhZTctNTliMC00MzgyLWE0ZGUtMDA3YTYyOWExMDBmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEwVDIzMDk0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1ZTY0NGM3NzNlNmQ0NmJiNDBiZjdhMGE1ZDM0OTdkMmIzMDQxMGMzZmIyMDM5MzkzZjg4NTBkZTE5MzYzZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cfX3e4X6WtrsfwooBhnQjFk1x6c7PVCbriBinmF9_Ns)

Ok, that explains it then. For reference this is what I get with batch_size 1, note it's iterations / second, so it can be even faster, not sure where the difference is from now though:

![Screenshot 2024-09-11 011700](https://github.com/user-attachments/assets/4c29f4ae-8057-417c-bf43-2cab9434bf89)

**Comment by plugcrypt** [comment]
Yeah, got 1.07s/it after a minute with AdamW, so for 20 images, I completed 1200 steps in about 20min. I still have to understand why increasing the batch size is increasing the training time, this should be the other way around right? Maybe I do need to divide the number of steps by the batch size value, I don't really know, cuz until now I used to set the number of repeats, the batch size and epoch to get the number of steps. Now the batch size 
seems to be automatically set insted of steps, I like that. 

So do I have to use gradiant type bf16, and save dtype as bf16 also? Or should I save dtype as fp8_e4m3fn in order to use it with a fp8 flux model, are there any incompatibilities we need to be aware of?

**Comment by plugcrypt** [comment]
The first one was saved in BF16 but was full of noise, the second one succeeded I guess
https://www.youtube.com/watch?v=8yrKXZJQPbE

**Comment by RaySteve312** [comment]
> For the same training settings on both trainer: 512x512 Adamw8bit 4 batch size
> 
> FluxGym : 1.19s/it ComfyUI FluxTrainer: 7.31s/it
> 
> About 6 times faster on FluxGym, on FluxTrainer my gpu utilisation is going up and down as well with my power draw, the vram usage doen't go above 55% with this training setting, I run this on a 4090
> 
> I hope that I'm missing something
> 
> Using these arguments : --use-pytorch-cross-attention --fast --highvram --dont-upcast-attention Eveything up to date, comfyUI + dependencies

in my case, only --highvram works. Do these command lines work in your training? my average speed in FluxTrain is quite slower than your speed, 4.5 s/it for batch size 1.

**Comment by RaySteve312** [comment]
> > No it's a mistake I made on my first post, But I previously compaired both with batch size 1 with the same settings and noticed a huge difference 6s/it on comfy, 1.2s/it on flux, but I didn't tried to compair this again with the split mode to false, and yes both are running at the same speed, this was the culprit. Thank you for the help appreciate
> > ![image](https://private-user-images.githubusercontent.com/88064700/366238064-758b1ae7-59b0-4382-a4de-007a629a100f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMTAwODUsIm5iZiI6MTcyNjAwOTc4NSwicGF0aCI6Ii84ODA2NDcwMC8zNjYyMzgwNjQtNzU4YjFhZTctNTliMC00MzgyLWE0ZGUtMDA3YTYyOWExMDBmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEwVDIzMDk0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1ZTY0NGM3NzNlNmQ0NmJiNDBiZjdhMGE1ZDM0OTdkMmIzMDQxMGMzZmIyMDM5MzkzZjg4NTBkZTE5MzYzZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cfX3e4X6WtrsfwooBhnQjFk1x6c7PVCbriBinmF9_Ns)
> 
> Ok, that explains it then. For reference this is what I get with batch_size 1, note it's iterations / second, so it can be even faster, not sure where the difference is from now though:
> 
> ![Screenshot 2024-09-11 011700](https://private-user-images.githubusercontent.com/40791699/366244365-4c29f4ae-8057-417c-bf43-2cab9434bf89.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNzcxNzcsIm5iZiI6MTcyNjA3Njg3NywicGF0aCI6Ii80MDc5MTY5OS8zNjYyNDQzNjUtNGMyOWY0YWUtODA1Ny00MTdjLWJmNDMtMmNhYjk0MzRiZjg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTExVDE3NDc1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2MWE0ZDNjZWMzN2I5ODMxZjk4M2ZhMWNlYTU1NzFkN2IwYzlhODU3NTY2NTYzZjM2NTRkN2RmM2JhMzFjZTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.87Efb9phdejsINDSe4Sz5FhdXxDKk9_PWOTPmpP8Sao)
you are so fast ! How can I speed up training?

**Comment by plugcrypt** [comment]
[CRT_LoRATraining.json](https://github.com/user-attachments/files/16969260/CRT_LoRATraining.json)

Trained two LoRA's with this, got around 1s/it at 512 batch 1, using adamw 

Still have to understand better the network dim and alpha, and do some tests, but this is working fine
The learning rate is pretty high but I guess It would gives better results on a low steps training

Cat Woman (Zoe Kravitz) [](https://civitai.com/models/739617/cat-woman-zoe-kravitz?modelVersionId=827119)
Helios 44-2 [](https://civitai.com/models/742089/helios-44-2?modelVersionId=829866)

**Comment by plugcrypt** [comment]
@kijai May I ask if it's possible to train a single transformer block?

**Comment by kijai** [comment]
> @kijai May I ask if it's possible to train a single transformer block?

Yes, it is. There's the block_args input to the unit node and block select node connects to that, for now it's just text about field but it has an example of the syntax as default.

**Comment by plugcrypt** [comment]
I have told that these ones could be affecting the style: 20, 22 - 7, 20 - 13, 15, and 25-27.
Have to do some testings on that, thank you +1

https://ostris.com/2024/09/07/skipping-flux-1-dev-blocks/

**Comment by RaySteve312** [comment]
> [CRT_LoRATraining.json](https://github.com/user-attachments/files/16969260/CRT_LoRATraining.json)
> 
> Trained two LoRA's with this, got around 1s/it at 512 batch 1, using adamw
> 
> Still have to understand better the network dim and alpha, and do some tests, but this is working fine The learning rate is pretty high but I guess It would gives better results on a low steps training
> 
> Cat Woman (Zoe Kravitz) https://civitai.com/models/739617/cat-woman-zoe-kravitz?modelVersionId=827119 Helios 44-2 https://civitai.com/models/742089/helios-44-2?modelVersionId=829866

Thanks! It means a lot for me! I will try it today!

**Comment by KLL535** [comment]
![Image](https://github.com/user-attachments/assets/f3da69ab-e435-48bc-859d-d5d57f1a3ade)

50 (!) times slower than FluxGum. Only 5 pictures. I don't know how to change the number of epoch, there is simply no such choice.

**Comment by PGCRT** [comment]
You are probably out of vram with a CPU fallback
try with batch size 1 
512x512px you don't need higher resolution 
set fp8_base on true, 

epochs are automatically defined based on your number of steps, number of imagesXrepeats and your batch size

if you want less epochs, increase the repeats of your dataset, but I recommend you to add more images instead, 5 is really low

**Comment by PGCRT** [comment]
You are probably out of vram with a CPU fallback
try with batch size 1
512x512px you don't need higher resolution
use an fp8 flux model
set fp8_base on true,

epochs are automatically defined based on your number of steps, number of imagesXrepeats and your batch size

if you want less epochs, increase the repeats of your dataset, but I recommend you to add more images instead, 5 is really low

**Comment by KLL535** [comment]
> You are probably out of vram with a CPU fallback try with batch size 1 512x512px you don't need higher resolution set fp8_base on true,
> 
> epochs are automatically defined based on your number of steps, number of imagesXrepeats and your batch size
> 
> if you want less epochs, increase the repeats of your dataset, but I recommend you to add more images instead, 5 is really low

batch size 1 - I tried that too, nothing changed
512x512px - that's true
fp8 flux model - that's true
fp8_base on true - that's true

Example workflow:
![Image](https://github.com/user-attachments/assets/d26e19c6-e68e-41a6-a6cf-fb492ad2678c)
I still don't understand how the epoch are calculated. In Fluxgum this was the basic setting, and here there are a bunch of settings, but epoch are not among them.

Does anyone have a working workflow for RTX3060 12 GB?

5 pictures for test only.
